adr/0018-public-ids-for-qualifications.md

```diff
@@ -0,0 +1,70 @@
+# 17. Application Choice Updated At
+
+Date: 2021-01-28
+
+## Status
+
+In progress
+
+## Context
+
+The Vendor API exposes candidate qualifications and grades for systems to ingest.
+Each qualification in the system has an id, which vendor systems use to identify the qualification.
+It is therefore important that each qualification coming from the API has a unique id.
+We use the database id for a qualification object for this purpose.
+
+An issue arises when we have qualifications that are made up of multiple exams/grades.
+See [the ADR on structured grades](./0016-structured-gcse-grades.md) for background on the way gcse qualifications are stored.
+API consumers expect one qualification entity per exam/grade.
+However, because of the structured grades introduced above, some exams/grades share the same database id.
+Therefore, not all the ids for qualifications coming through the API are unique.
+
+## Decision
+
+We initially considered changing the data type of the id field in the API response to a string, to support ids of the form `123`, `123_1`, `123_2`.
+Vendors were not able to support that change, so the field had to remain an integer.
+
+The solution for this will be split into two: one for qualifications with a simple grade, and one for qualifications with structured grades.
+The application_qualifications table will get a new public_id column to store the API id for a qualification.
+For qualifications with simple grades (only one exam/grade), the public_id will just be set to the database id of the qualification initially via a data migration.
+For qualifications with structured grades (built from multiple exams/grades), the public_id column will be set to `nil`, and the public id will be set inside the structured_grades json, with one id per grade.
+A qualification with structured grades could look like this:
+```
+{
+    id: 123
+    subject: 'english',
+    public_id: nil,
+    grade: nil,
+    structured_grades: {
+      'English Language': { grade: 'E', public_id: '123' },
+      'English Literature': { grade: 'B', public_id: '1093' },
+      'Cockney Rhyming Slang': { grade: 'A*', public_id: '1094' }
+    },
+    level: 'gcse'
+}
+```
```

Comment:
It is possible to edit qualifications between structured and simple grade ones. E.g. I can start with
```
{
    id: 123
    subject: 'english',
    public_id: 1092,
    grade: 'A',
    structured_grades: nil,
    level: 'gcse'
}
```
Then edit it to be a multiple part GCSE:
```
{
    id: 123
    subject: 'english',
    public_id: nil,
    grade: nil,
    structured_grades: {
      'English Language': { grade: 'E', public_id: '123' },
      'English Literature': { grade: 'B', public_id: '1093' },
      'Cockney Rhyming Slang': { grade: 'A*', public_id: '1094' }
    },
    level: 'gcse'
}
```
Then what happens if I change my mind back again?
```
{
    id: 123
    subject: 'english',
    public_id: ????,
    grade: 'A',
    structured_grades: nil,
    level: 'gcse'
}
```
This isn't really a new qualification object, so giving it a new public_id doesn't seem correct.

Do we want to try to remember the old public_id? Is this even an issue?

Reply:
From an API point of view you'll change the set from 

`[ { id: 123...}, { id: 1093 ...}, { id: 1094...} ]`

to 

`[{ id: 123...}]`

I don't know how SRSs will cope will removed IDs — hopefully they'll tolerate them.

In practice, though, this will hardly ever happen because providers don't see applications until they're submitted, so the candidate should get it right first. If they did need to change after submission we'd handle the change through support and probably advise the provider too as we'd assume they'd already looked at it.

Reply:
To answer the initial question, this can't happen with English GCSEs, because they are always stored as structured grades. But it could happen with science GCSEs where single and double awards are stored _normally_, but triple awards are structured as json.

I agree with Duncan on the next part. This should be set once the applications are submitted.